packages/react-server/core/renderMiddleware.js

```diff
@@ -105,16 +105,25 @@ module.exports = function(server, routes) {
 			navigateDfd.resolve();
 
 			if (err) {
-				logger.log("onNavigate received a non-2xx HTTP code", err);
-				handleResponseComplete(req, res, context, start, page);
-				if (err.status && err.status === 404) {
-					next();
-				} else if (err.status === 301 || err.status === 302 || err.status === 307) {
+				// The page can elect to proceed to render
+				// even with a non-2xx response.  If it
+				// _doesn't_ do so then we're done.
+				var done = !(page && page.getHaveDocument());
+
+				if (err.status === 301 || err.status === 302 || err.status === 307) {
 					res.redirect(err.status, err.redirectUrl);
-				} else {
-					next(err);
+				} else if (done) {
```

Comment:
Sorry, yeah, I was unclear in what I meant. I think `res.redirect` adds the appropriate body -- the target URL for the redirect. I don't think there's ever a case where we'd want to fall through _past_ the `if (done)` call below, and it's hard to tell whether or not that's the case just by looking at this code. (Though it probably is the case? Unless somebody says `setHaveDocument` and then sends a redirect?).

I wonder if just saying `done = true` right under the `res.redirect` call would be clearer.